File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.java
import java.util.*;

class Solution {
   public int solution(String[] friends, String[] gifts) {

        HashMap<String, Integer> giftPoint = new HashMap<>();
        HashMap<String ,Integer> indexMap  = new HashMap<>();
        int[][] graph = new int[friends.length][friends.length];

        int maxGift = 0;

        // 인덱스 저장
        for (int i = 0; i < friends.length; i++) {
            indexMap.put(friends[i], i);
            giftPoint.put(friends[i], 0);
        }
       
        // 선물 점수 기록, 그래프 기록
        for (int i = 0; i < gifts.length; i++) {
            String[] giveTake = gifts[i].split(" ");
            String giver = giveTake[0];
            String taker = giveTake[1];

            giftPoint.put(giver, giftPoint.get(giver)+1);
            giftPoint.put(taker, giftPoint.get(taker)-1);
            graph[indexMap.get(giver)][indexMap.get(taker)] += 1;
        }

        for (int i = 0; i < friends.length; i++) {
            int cur = 0;
            
            for (int j = 0; j < friends.length; j++) {
                if (i == j) {
                    continue;
                }
                
                if (graph[i][j] > graph[j][i]) {
                    cur += 1;
                } else if (graph[i][j] == graph[j][i] && giftPoint.get(friends[i]) > giftPoint.get(friends[j])) {
                    cur += 1;
                }
            }

            maxGift = Math.max(maxGift, cur);
        }
       
//         for (int i = 0; i < friends.length; i++) {
//             System.out.println(Arrays.toString(graph[i]));
//         }
       
//         for (int i = 0; i < friends.length; i++) {
//             System.out.println(giftPoint.get(friends[i]));
//         }
        
        return maxGift;
    }
}

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.py
def solution(friends, gifts):
    sequence = {}
    graph = [[0]*len(friends) for i in range(len(friends))]
    gift = {}
    check = []
    for i, n in enumerate(friends):
        sequence[n] = i
        gift[n] = [0,0,0]
    for i in gifts:
        k = i.split()
        give = k[0]
        get = k[1]
        gift[give][0] += 1
        gift[get][1] += 1
        gift[give][2] = gift[give][0] - gift[give][1]
        gift[get][2] = gift[get][0] - gift[get][1]
        graph[sequence[give]][sequence[get]] += 1
    for i in range(len(graph)):
        get_gift = 0
        for j in range(len(graph)):
            if i == j:
                continue
            if graph[i][j] > graph[j][i]:
                get_gift += 1
            elif graph[i][j] == graph[j][i] and gift[friends[i]][2] > gift[friends[j]][2]:
                get_gift+=1
        check.append(get_gift)
    return max(check)
